src/components/accession2/withdraw/WeightWithdrawal.tsx

```diff
@@ -0,0 +1,241 @@
+    let estimated = 0;
+    let valid = false;
+
+    if (
+      value.toString() === accession.remainingQuantity?.quantity.toString() &&
+      withdrawnQuantity?.units === accession.remainingQuantity?.units
+    ) {
+      setWithdrawAllSelected(true);
+    } else {
+      setWithdrawAllSelected(false);
+    }
+
+    if (accession.subsetCount && accession.subsetWeight) {
+      if (
+        accession.remainingQuantity?.units &&
+        accession.remainingQuantity?.units === 'Seeds' &&
+        accession.estimatedWeight?.units
+      ) {
+        estimated =
+          convertUnits(value, accession.estimatedWeight?.units, accession.subsetWeight.units) *
+          (accession.subsetCount / accession.subsetWeight.quantity);
+      } else if (accession.remainingQuantity?.units) {
+        estimated =
+          convertUnits(value, accession.remainingQuantity?.units, accession.subsetWeight.units) *
+          (accession.subsetCount / accession.subsetWeight.quantity);
+      }
+      setEstimatedWithdrawnCt(estimated);
+    }
+
+    setWithdrawnQuantity(
+      value.toString().trim() === ''
+        ? undefined
+        : {
+            quantity: value || 0,
+            units: withdrawnQuantity?.units || accession.remainingQuantity?.units || 'Grams',
+          }
+    );
+
+    setWithdrawnQtyError('');
+    if (purpose === 'Nursery' || purpose === 'Viability Testing') {
+      if (!accession.estimatedCount || !accession.subsetWeight?.quantity || !accession.subsetCount) {
+        setWithdrawnQtyError(
+          purpose === 'Nursery'
+            ? strings.MISSING_SUBSET_WEIGHT_ERROR_NURSERY
+            : strings.MISSING_SUBSET_WEIGHT_ERROR_VIABILITY_TEST
+        );
+      }
+    }
+    valid = validateAmount(estimated, value, withdrawnQuantity?.units || accession.remainingQuantity?.units || 'Grams');
+    if (purpose === 'Nursery' || purpose === 'Viability Testing') {
+      onWithdrawCtUpdate({ quantity: estimated, units: 'Seeds' }, valid);
+    } else {
+      onWithdrawCtUpdate(
+        {
+          quantity: value || 0,
+          units: withdrawnQuantity?.units || accession.remainingQuantity?.units || 'Grams',
+        },
+        valid
+      );
+    }
+  };
+
+  const onSelectAll = (id: string, withdrawAll: boolean) => {
+    if (withdrawAll) {
+      if (
+        accession.remainingQuantity?.units &&
+        accession.remainingQuantity?.units === 'Seeds' &&
+        accession.estimatedWeight?.quantity
+      ) {
+        onChangeWithdrawnQuantity(accession.estimatedWeight?.quantity);
+      } else if (accession.remainingQuantity?.units && accession.remainingQuantity?.units !== 'Seeds') {
+        onChangeWithdrawnQuantity(accession.remainingQuantity?.quantity);
+      }
+    } else {
+      setWithdrawnQuantity(undefined);
+    }
+
+    setWithdrawAllSelected(withdrawAll);
+  };
+
+  const validateAmount = (estimated: number, withdrawnQty: number, withdrawnUnits: string) => {
```

Comment:
This function could benefit from "never nesting". Not really necessary, but would help to improve the readability. Basically, try to only have 1 level of nesting for each validation case. For example:

```
if (!withdrawnQty) {
  setWithdrawnQtyError(strings.REQUIRED_FIELD);
  return false;
}

if (isNaN(withdrawnQty) || Number(withdrawnQty) <= 0) {
  setWithdrawnQtyError(strings.INVALID_VALUE);
  return false;
}
...
```

Reply:
Love this! I also aim to limit nesting to one level in code to improve readability.

Reply:
This was a good exercise.  Makes for some duplicate code, but definitely more readable.